src/decode.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use crate::{Expect, Index, Mir, FORMAT_VERSION};
+use rmp_serde::{
+    decode::{self, ReadReader},
+    Deserializer,
+};
+use serde::Deserialize;
+use std::io::Read;
+
+/// A decoder.
+pub struct Decoder<'a> {
+    deser: Deserializer<ReadReader<&'a mut dyn Read>>,
+    index: Index,
+}
+
+impl<'a> Decoder<'a> {
+    /// Returns a new decoder which will deserialise from `read_from` and also an `Index` informing
+    /// the consumer of the expected contents.
```

Comment:
I wonder if this docstring should mention that the ABI version number is checked.

(Sorry, only thought of this after raising the PR).

Reply:
Presumably that's not the only check that, conceptually, it might one day do?

Reply:
It's not impossible that we might check something else at some point, but I can't think of anything off the top of my head.

What are you suggesting? That it's not worth mentioning?

Reply:
Yeah, I doubt it's worth mentioning.